File: shoebill_libs/DatabaseManager/src/main/java/lt/ltrp/DatabasePlugin.java
package lt.ltrp;

import com.mchange.v2.c3p0.ComboPooledDataSource;
import net.gtaun.shoebill.resource.Plugin;
import org.slf4j.Logger;

import javax.sql.DataSource;
import java.beans.PropertyVetoException;
import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.Properties;

/**
 * @author Bebras
 *         2016.03.02.
 */
public class DatabasePlugin extends Plugin {

    private static final String driver = "com.mysql.jdbc.Driver";
    private static final String user = "root";
    private static final String password = "";
    private static final String database = "ltrp-java";
    private static final String DEFAULT_CHARSET = "utf8";
    private static final String DEFAULT_COLLATE = "utf8_unicode_ci";
    private static final String url = "jdbc:mysql://localhost:3306/" + database + "?characterEncoding=" + DEFAULT_CHARSET + "";

    private ComboPooledDataSource dataSource;

    @Override
    protected void onEnable() throws Throwable {
        Logger logger = getLogger();
        logger.debug("onEnable");
        dataSource = new ComboPooledDataSource();
        try {
            dataSource.setDriverClass(driver);
        } catch (PropertyVetoException e) {
            e.printStackTrace();
        }
        Properties props = dataSource.getProperties();
        props.put("com.mchange.v2.log.MLog", "com.mchange.v2.log.slf4j.Slf4jMLog");
        props.put("com.mchange.v2.log.NameTransformer", "com.mchange.v2.log.PackageNames");
        props.put("charSet", DEFAULT_CHARSET);
        dataSource.setProperties(props);
        dataSource.setJdbcUrl(url);
        dataSource.setUser(user);
        dataSource.setPassword(password);

        dataSource.setMinPoolSize(1);
        dataSource.setMaxPoolSize(6);

        if(dataSource == null) {
            throw new RuntimeException("Datasource is null");
        }
        try (
                Connection connection = dataSource.getConnection();
                Statement stmt = connection.createStatement();
        ) {
            stmt.executeUpdate("SET NAMES " + DEFAULT_CHARSET);
            stmt.executeUpdate("SET character_set_client=" + DEFAULT_CHARSET);
            stmt.executeUpdate("SET character_set_results=" + DEFAULT_CHARSET);
            stmt.executeUpdate("SET character_set_connection=" + DEFAULT_CHARSET);
            stmt.executeUpdate("SET CHARACTER SET " + DEFAULT_CHARSET);
            stmt.executeUpdate("SET character_set_database=" + DEFAULT_CHARSET);
            stmt.executeUpdate("SET character_set_server="+DEFAULT_CHARSET);

            ResultSet r = stmt.executeQuery("show variables like '%character%'");
            while(r.next()) {
                logger.debug(r.getObject(1) + "=" + r.getObject(2));
            }
        } catch (SQLException e) {
            logger.error(getClass().getName() + " Could not set charset");
        }

        logger.info("Database plugin loaded. DataSource:" + dataSource.toString());
    }

    public DataSource getDataSource() {
        return dataSource;
    }


    @Override
    protected void onDisable() throws Throwable {
        dataSource.close();
    }
}

File: shoebill_libs/DatabaseManager/src/main/java/lt/maze/DatabasePlugin.kt
package lt.maze

import com.mchange.v2.c3p0.ComboPooledDataSource
import net.gtaun.shoebill.ShoebillMain
import net.gtaun.shoebill.resource.Plugin
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.sql.Connection
import java.sql.SQLException
import java.sql.Statement

/**
 * Created by Bebras on 2016-12-24.
 * A plugin for opening a pooled connection to a sql database
 */

@ShoebillMain(name = "Database manager plugin",
        author = "Bebras",
        description = "Creates a datasource, thats all this plugin deoes.",
        version = "1.1")
class DatabasePlugin : Plugin() {

    lateinit var dataSource: ComboPooledDataSource

    override fun onEnable() {
        val logger = logger
        val ds = ComboPooledDataSource()
        try {
            ds.driverClass = driver
        } catch (e: Exception) {
            logger.error("Setting driver to \"" + driver + "\" failed. " + e.message, e.cause)
        }

        val props = ds.properties
        props.put("com.mchange.v2.log.MLog", "com.mchange.v2.log.slf4j.Slf4jMLog")
        props.put("com.mchange.v2.log.NameTransformer", "com.mchange.v2.log.PackageNames")
        props.put("charSet", DEFAULT_CHARSET)
        ds.properties = props
        ds.jdbcUrl = url
        ds.user = user
        ds.password = password
        ds.minPoolSize = 1
        ds.maxPoolSize = 6

        var connection: Connection? = null
        var stmt: Statement? = null
        try {
            connection = ds.connection
            stmt = connection.createStatement()
            stmt.executeUpdate("SET NAMES " + DEFAULT_CHARSET)
            stmt.executeUpdate("SET character_set_client=" + DEFAULT_CHARSET)
            stmt.executeUpdate("SET character_set_results=" + DEFAULT_CHARSET)
            stmt.executeUpdate("SET character_set_connection=" + DEFAULT_CHARSET)
            stmt.executeUpdate("SET CHARACTER SET " + DEFAULT_CHARSET)
            stmt.executeUpdate("SET character_set_database=" + DEFAULT_CHARSET)
            stmt.executeUpdate("SET character_set_server=" + DEFAULT_CHARSET)

            val r = stmt.executeQuery("show variables like '%character%'")
            while (r.next()) {
                logger.debug(r.getObject(1).toString() + "=" + r.getObject(2))
            }
            logger.info("Charset successfully set to " + DEFAULT_CHARSET)
        } catch (e: SQLException) {
            logger.error("Could not set charset")
        } finally {
            connection?.close()
            stmt?.close()
        }

        dataSource = ds
        logger.info("Database plugin loaded. DataSource:" + dataSource.toString())
    }

    override fun onDisable() {
        logger.info("Shutting down..")
        dataSource.close()
    }

    @Throws(SQLException::class)
    fun query(query: String): Boolean {
        var connection: Connection? = null
        var stmt: Statement? = null
        try {
            connection = dataSource.connection
            stmt = connection.createStatement()
            return stmt.execute(query)
        } finally {
            connection?.close()
            stmt?.close()
        }

    }

    @Throws(SQLException::class, IOException::class)
    fun query(inputStream: InputStream): Boolean {
        val reader = BufferedReader(InputStreamReader(inputStream))
        reader.use {
            return query(reader.readText())
        }
    }

    companion object {
        private val driver = "com.mysql.jdbc.Driver"
        private val user = "root"
        private val password = ""
        private val database = "ltrp-java"
        private val DEFAULT_CHARSET = "utf8"
        private val DEFAULT_COLLATE = "utf8_unicode_ci"
        private val url = "jdbc:mysql://localhost:3306/$database?characterEncoding=$DEFAULT_CHARSET"

    }
}